// @ai-hero/sandcastle@0.4.0/dist/AgentProvider.js
const shellEscape = (s) => "'" + s.replace(/'/g, "'\\''") + "'";
/** Maps allowlisted tool names to the input field containing the display arg */
const TOOL_ARG_FIELDS = {
    Bash: "command",
    WebSearch: "query",
    WebFetch: "url",
    Agent: "description",
};
const parseStreamJsonLine = (line) => {
    if (!line.startsWith("{"))
        return [];
    try {
        const obj = JSON.parse(line);
        if (obj.type === "assistant" && Array.isArray(obj.message?.content)) {
            const events = [];
            const texts = [];
            for (const block of obj.message.content) {
                if (block.type === "text" && typeof block.text === "string") {
                    texts.push(block.text);
                }
                else if (block.type === "tool_use" &&
                    typeof block.name === "string" &&
                    block.input !== undefined) {
                    const argField = TOOL_ARG_FIELDS[block.name];
                    if (argField === undefined)
                        continue; // not allowlisted
                    const argValue = block.input[argField];
                    if (typeof argValue !== "string")
                        continue; // missing/wrong arg field
                    if (texts.length > 0) {
                        events.push({ type: "text", text: texts.join("") });
                        texts.length = 0;
                    }
                    events.push({
                        type: "tool_call",
                        name: block.name,
                        args: argValue,
                    });
                }
            }
            if (texts.length > 0) {
                events.push({ type: "text", text: texts.join("") });
            }
            return events;
        }
        if (obj.type === "result" && typeof obj.result === "string") {
            return [{ type: "result", result: obj.result }];
        }
    }
    catch {
        // Not valid JSON — skip
    }
    return [];
};
export const DEFAULT_MODEL = "claude-opus-4-6";
// ---------------------------------------------------------------------------
// Pi agent provider
// ---------------------------------------------------------------------------
const parsePiStreamLine = (line) => {
    if (!line.startsWith("{"))
        return [];
    try {
        const obj = JSON.parse(line);
        if (obj.type === "message_update" && obj.assistantMessageEvent) {
            const evt = obj.assistantMessageEvent;
            if (evt.type === "text_delta" && typeof evt.delta === "string") {
                return [{ type: "text", text: evt.delta }];
            }
            return [];
        }
        if (obj.type === "tool_execution_start") {
            const toolName = obj.toolName;
            if (typeof toolName !== "string")
                return [];
            const argField = TOOL_ARG_FIELDS[toolName];
            if (argField === undefined)
                return [];
            const args = obj.args;
            if (!args)
                return [];
            const argValue = args[argField];
            if (typeof argValue !== "string")
                return [];
            return [{ type: "tool_call", name: toolName, args: argValue }];
        }
        if (obj.type === "agent_end" && Array.isArray(obj.messages)) {
            const messages = obj.messages;
            for (let i = messages.length - 1; i >= 0; i--) {
                const msg = messages[i];
                if (msg?.role === "assistant") {
                    const texts = [];
                    for (const block of msg.content) {
                        if (block.type === "text" && typeof block.text === "string") {
                            texts.push(block.text);
                        }
                    }
                    if (texts.length > 0) {
                        return [{ type: "result", result: texts.join("") }];
                    }
                    break;
                }
            }
            return [];
        }
    }
    catch {
        // Not valid JSON — skip
    }
    return [];
};
export const pi = (model) => ({
    name: "pi",
    buildPrintCommand(prompt) {
        return `pi -p --mode json --no-session --model ${shellEscape(model)} ${shellEscape(prompt)}`;
    },
    buildInteractiveArgs(_prompt) {
        return ["pi", "--model", model];
    },
    parseStreamLine(line) {
        return parsePiStreamLine(line);
    },
});
// ---------------------------------------------------------------------------
// Codex agent provider
// ---------------------------------------------------------------------------
const parseCodexStreamLine = (line) => {
    if (!line.startsWith("{"))
        return [];
    try {
        const obj = JSON.parse(line);
        // item.completed with agent_message → text + result
        if (obj.type === "item.completed" &&
            obj.item?.type === "agent_message" &&
            typeof obj.item.content === "string") {
            const text = obj.item.content;
            return [
                { type: "text", text },
                { type: "result", result: text },
            ];
        }
        // item.started with command_execution → tool call
        if (obj.type === "item.started" &&
            obj.item?.type === "command_execution" &&
            typeof obj.item.command === "string") {
            return [{ type: "tool_call", name: "Bash", args: obj.item.command }];
        }
        // turn.completed → skip
    }
    catch {
        // Not valid JSON — skip
    }
    return [];
};
export const codex = (model) => ({
    name: "codex",
    buildPrintCommand(prompt) {
        return `codex exec --json --dangerously-bypass-approvals-and-sandbox -m ${shellEscape(model)} ${shellEscape(prompt)}`;
    },
    buildInteractiveArgs(_prompt) {
        return ["codex", "--model", model];
    },
    parseStreamLine(line) {
        return parseCodexStreamLine(line);
    },
});
export const claudeCode = (model, options) => ({
    name: "claude-code",
    buildPrintCommand(prompt) {
        const effortFlag = options?.effort ? ` --effort ${options.effort}` : "";
        return `claude --print --verbose --dangerously-skip-permissions --output-format stream-json --model ${shellEscape(model)}${effortFlag} -p ${shellEscape(prompt)}`;
    },
    buildInteractiveArgs(_prompt) {
        const args = ["claude", "--dangerously-skip-permissions", "--model", model];
        if (options?.effort)
            args.push("--effort", options.effort);
        return args;
    },
    parseStreamLine(line) {
        return parseStreamJsonLine(line);
    },
});
