// @ai-hero/sandcastle@0.4.0/dist/createSandbox.js
import { NodeContext, NodeFileSystem } from "@effect/platform-node";
import { join } from "node:path";
import { Effect, Layer, Ref } from "effect";
import { Display, FileDisplay, SilentDisplay, } from "./Display.js";
import { resolveEnv } from "./EnvResolver.js";
import { orchestrate } from "./Orchestrator.js";
import { substitutePromptArgs, validateNoBuiltInArgOverride, BUILT_IN_PROMPT_ARG_KEYS, } from "./PromptArgumentSubstitution.js";
import { resolvePrompt } from "./PromptResolver.js";
import { buildLogFilename, printFileDisplayStartup } from "./run.js";
import { Sandbox as SandboxTag, SandboxFactory, SANDBOX_WORKSPACE_DIR, makeSandboxLayerFromHandle, resolveGitMounts, } from "./SandboxFactory.js";
import * as WorktreeManager from "./WorktreeManager.js";
import { copyToSandbox } from "./CopyToSandbox.js";
/**
 * Eagerly creates a git worktree on the provided explicit branch and starts
 * a Docker container (or local sandbox in test mode) with the worktree
 * bind-mounted. Returns a Sandbox handle that can be reused across multiple
 * `run()` calls.
 */
export const createSandbox = async (options) => {
    const hostRepoDir = options._test?.hostRepoDir ?? process.cwd();
    const { branch } = options;
    const isTestMode = !!options._test?.buildSandboxLayer;
    // 1. Prune stale worktrees + create worktree on the explicit branch
    const worktreeInfo = await Effect.runPromise(WorktreeManager.pruneStale(hostRepoDir)
        .pipe(Effect.catchAll(() => Effect.void))
        .pipe(Effect.andThen(WorktreeManager.create(hostRepoDir, { branch })))
        .pipe(Effect.provide(NodeContext.layer)));
    const worktreePath = worktreeInfo.path;
    // 2. Copy files if requested
    if (options.copyToSandbox && options.copyToSandbox.length > 0) {
        await Effect.runPromise(copyToSandbox(options.copyToSandbox, hostRepoDir, worktreePath));
    }
    // 3. Start sandbox via provider or local sandbox layer (test mode)
    let providerHandle;
    let sandboxLayer;
    let sandboxRepoDir;
    if (isTestMode) {
        sandboxLayer = options._test.buildSandboxLayer(worktreePath);
        sandboxRepoDir = worktreePath;
    }
    else {
        // Provider mode: delegate to the sandbox provider
        const env = await Effect.runPromise(resolveEnv(hostRepoDir).pipe(Effect.provide(NodeContext.layer)));
        const gitPath = join(hostRepoDir, ".git");
        const gitMounts = await Effect.runPromise(resolveGitMounts(gitPath).pipe(Effect.provide(NodeFileSystem.layer)));
        const mounts = [
            { hostPath: worktreePath, sandboxPath: SANDBOX_WORKSPACE_DIR },
            ...gitMounts,
        ];
        providerHandle = await options.sandbox.create({
            worktreePath,
            hostRepoPath: hostRepoDir,
            mounts,
            env,
        });
        sandboxLayer = makeSandboxLayerFromHandle(providerHandle);
        sandboxRepoDir = providerHandle.workspacePath;
    }
    // 4. Run onSandboxReady hooks
    if (options.hooks?.onSandboxReady?.length) {
        await Effect.runPromise(Effect.gen(function* () {
            const sandbox = yield* SandboxTag;
            yield* sandbox.exec(`git config --global --add safe.directory "${sandboxRepoDir}"`);
            for (const hook of options.hooks.onSandboxReady) {
                yield* sandbox.exec(hook.command, { cwd: sandboxRepoDir });
            }
        }).pipe(Effect.provide(sandboxLayer)));
    }
    // 5. Set up signal handlers
    let closed = false;
    const forceCleanup = () => {
        console.error(`\nWorktree preserved at ${worktreePath}`);
        console.error(`  To review: cd ${worktreePath}`);
        console.error(`  To clean up: git worktree remove --force ${worktreePath}`);
    };
    const onSignal = () => {
        forceCleanup();
        process.exit(1);
    };
    process.on("SIGINT", onSignal);
    process.on("SIGTERM", onSignal);
    // 6. Build close function
    const doClose = async () => {
        if (closed)
            return { preservedWorktreePath: undefined };
        closed = true;
        // Close provider handle
        if (providerHandle) {
            await providerHandle.close();
        }
        // Check for uncommitted changes
        const isDirty = await Effect.runPromise(WorktreeManager.hasUncommittedChanges(worktreePath).pipe(Effect.catchAll(() => Effect.succeed(false))));
        if (isDirty) {
            return { preservedWorktreePath: worktreePath };
        }
        // Remove worktree
        await Effect.runPromise(WorktreeManager.remove(worktreePath).pipe(Effect.catchAll(() => Effect.void)));
        return { preservedWorktreePath: undefined };
    };
    // 7. Return the Sandbox handle
    const sandboxHandle = {
        branch,
        worktreePath,
        run: async (runOptions) => {
            const { agent: provider, prompt, promptFile, maxIterations = 1, } = runOptions;
            // Resolve prompt
            const rawPrompt = await Effect.runPromise(resolvePrompt({ prompt, promptFile }).pipe(Effect.provide(NodeContext.layer)));
            // Resolve prompt arguments
            const userArgs = runOptions.promptArgs ?? {};
            const currentHostBranch = await Effect.runPromise(WorktreeManager.getCurrentBranch(hostRepoDir));
            const displayRef = Ref.unsafeMake([]);
            const silentDisplayLayer = SilentDisplay.layer(displayRef);
            const resolvedPrompt = await Effect.runPromise(Effect.gen(function* () {
                yield* validateNoBuiltInArgOverride(userArgs);
                const effectiveArgs = {
                    SOURCE_BRANCH: branch,
                    TARGET_BRANCH: currentHostBranch,
                    ...userArgs,
                };
                const builtInArgKeysSet = new Set(BUILT_IN_PROMPT_ARG_KEYS);
                return yield* substitutePromptArgs(rawPrompt, effectiveArgs, builtInArgKeysSet);
            }).pipe(Effect.provide(silentDisplayLayer)));
            // Resolve logging
            const resolvedLogging = runOptions.logging ?? {
                type: "file",
                path: join(hostRepoDir, ".sandcastle", "logs", buildLogFilename(branch, undefined, runOptions.name)),
            };
            const runDisplayLayer = resolvedLogging.type === "file"
                ? (() => {
                    printFileDisplayStartup({
                        logPath: resolvedLogging.path,
                        agentName: runOptions.name,
                        branch,
                    });
                    return Layer.provide(FileDisplay.layer(resolvedLogging.path), NodeFileSystem.layer);
                })()
                : silentDisplayLayer;
            // Build a SandboxFactory that reuses the existing container/sandbox
            const reuseFactoryLayer = Layer.succeed(SandboxFactory, {
                withSandbox: (makeEffect) => makeEffect({ hostWorktreePath: worktreePath }).pipe(Effect.provide(sandboxLayer), Effect.map((value) => ({
                    value,
                    preservedWorktreePath: undefined,
                }))),
            });
            const runLayer = Layer.merge(reuseFactoryLayer, runDisplayLayer);
            const result = await Effect.runPromise(Effect.gen(function* () {
                const display = yield* Display;
                yield* display.intro(runOptions.name ?? "sandcastle");
                return yield* orchestrate({
                    hostRepoDir,
                    sandboxRepoDir,
                    iterations: maxIterations,
                    prompt: resolvedPrompt,
                    branch,
                    provider,
                    completionSignal: runOptions.completionSignal,
                    idleTimeoutSeconds: runOptions.idleTimeoutSeconds,
                    name: runOptions.name,
                });
            }).pipe(Effect.provide(runLayer)));
            return {
                iterationsRun: result.iterationsRun,
                completionSignal: result.completionSignal,
                stdout: result.stdout,
                commits: result.commits,
                logFilePath: resolvedLogging.type === "file" ? resolvedLogging.path : undefined,
            };
        },
        close: async () => {
            process.removeListener("SIGINT", onSignal);
            process.removeListener("SIGTERM", onSignal);
            return doClose();
        },
        [Symbol.asyncDispose]: async () => {
            await sandboxHandle.close();
        },
    };
    return sandboxHandle;
};
