// @ai-hero/sandcastle@0.4.0/dist/DockerLifecycle.js
import { Effect } from "effect";
import { execFile } from "node:child_process";
import { resolve } from "node:path";
import { DockerError } from "./errors.js";
const dockerExec = (args) => Effect.async((resume) => {
    execFile("docker", args, { maxBuffer: 10 * 1024 * 1024 }, (error, stdout, stderr) => {
        if (error) {
            resume(Effect.fail(new DockerError({
                message: `docker ${args[0]} failed: ${stderr?.toString() || error.message}`,
            })));
        }
        else {
            resume(Effect.succeed(stdout.toString()));
        }
    });
});
/**
 * Build the sandcastle Docker image.
 *
 * When `dockerfile` is provided, uses `docker build -f <dockerfile> <cwd>`
 * so COPY instructions resolve relative to the current working directory.
 * Otherwise, uses `docker build <dockerfileDir>` (the default .sandcastle/ directory).
 */
export const buildImage = (imageName, dockerfileDir, options) => Effect.gen(function* () {
    if (options?.dockerfile) {
        yield* dockerExec([
            "build",
            "-t",
            imageName,
            "-f",
            resolve(options.dockerfile),
            process.cwd(),
        ]);
    }
    else {
        yield* dockerExec(["build", "-t", imageName, resolve(dockerfileDir)]);
    }
});
/**
 * Start a new container with environment variables injected.
 */
export const startContainer = (containerName, imageName, env, options) => Effect.gen(function* () {
    // Check if container already exists
    const existing = yield* dockerExec([
        "ps",
        "-a",
        "--filter",
        `name=^${containerName}$`,
        "--format",
        "{{.Names}}",
    ]);
    if (existing.trim() === containerName) {
        yield* Effect.fail(new DockerError({
            message: `Container '${containerName}' already exists. Run cleanup first.`,
        }));
    }
    const envFlags = Object.entries(env).flatMap(([k, v]) => [
        "-e",
        `${k}=${v}`,
    ]);
    const volumeFlags = (options?.volumeMounts ?? []).flatMap((mount) => [
        "-v",
        mount,
    ]);
    const workdirFlags = options?.workdir ? ["-w", options.workdir] : [];
    const userFlags = options?.user ? ["--user", options.user] : [];
    yield* dockerExec([
        "run",
        "-d",
        "--name",
        containerName,
        ...envFlags,
        ...volumeFlags,
        ...workdirFlags,
        ...userFlags,
        imageName,
    ]);
});
/**
 * Fix ownership of a directory inside the container.
 * Runs as root so the target owner can write to the path.
 *
 * Non-fatal: if chown fails (e.g. read-only .git/objects on macOS VirtioFS),
 * a warning is logged but the error is not propagated.
 *
 * @param owner - chown-compatible owner spec, e.g. "1000:1000" or "agent"
 */
export const chownInContainer = (containerName, owner, path) => Effect.asVoid(dockerExec([
    "exec",
    "-u",
    "root",
    containerName,
    "chown",
    "-R",
    owner,
    path,
])).pipe(Effect.catchAll((error) => {
    console.warn(`chown -R ${owner} ${path} in container ${containerName} failed (non-fatal): ${error.message}`);
    return Effect.void;
}));
/**
 * Stop and remove a container without removing the image.
 */
export const removeContainer = (containerName) => Effect.gen(function* () {
    // Stop container (ignore errors if already stopped)
    yield* Effect.ignore(dockerExec(["stop", containerName]));
    // Remove container (ignore errors if not found)
    yield* Effect.ignore(dockerExec(["rm", containerName]));
});
/**
 * Remove a Docker image.
 */
export const removeImage = (imageName) => Effect.gen(function* () {
    yield* dockerExec(["rmi", imageName]);
});
