// @ai-hero/sandcastle@0.4.0/dist/errors.js
import { Data } from "effect";
/** Command execution failed in the sandbox */
export class ExecError extends Data.TaggedError("ExecError") {
}
/** Command execution failed on the host */
export class ExecHostError extends Data.TaggedError("ExecHostError") {
}
/** File copy between host and sandbox failed */
export class CopyError extends Data.TaggedError("CopyError") {
}
/** Docker infrastructure operation failed */
export class DockerError extends Data.TaggedError("DockerError") {
}
/** Git sync-in or sync-out operation failed */
export class SyncError extends Data.TaggedError("SyncError") {
}
/** Git worktree operation failed */
export class WorktreeError extends Data.TaggedError("WorktreeError") {
}
/** Prompt resolution or preprocessing failed */
export class PromptError extends Data.TaggedError("PromptError") {
}
/** Agent invocation failed */
export class AgentError extends Data.TaggedError("AgentError") {
}
/** .sandcastle/ config directory missing */
export class ConfigDirError extends Data.TaggedError("ConfigDirError") {
}
/** Initialization or setup operation failed */
export class InitError extends Data.TaggedError("InitError") {
}
/** Run exceeded the configured idle timeout */
export class TimeoutError extends Data.TaggedError("TimeoutError") {
}
