// @ai-hero/sandcastle@0.4.0/dist/Orchestrator.js
import { Deferred, Effect } from "effect";
import { Display } from "./Display.js";
import { preprocessPrompt } from "./PromptPreprocessor.js";
import { AgentError, TimeoutError } from "./errors.js";
import { SandboxFactory } from "./SandboxFactory.js";
import { withSandboxLifecycle } from "./SandboxLifecycle.js";
const IDLE_WARNING_INTERVAL_MS = 60_000;
const invokeAgent = (sandbox, sandboxRepoDir, prompt, provider, idleTimeoutMs, onText, onToolCall, onIdleWarning, idleWarningIntervalMs = IDLE_WARNING_INTERVAL_MS) => Effect.gen(function* () {
    let resultText = "";
    // Deferred that will be failed when the idle timer fires
    const timeoutSignal = yield* Deferred.make();
    let timeoutHandle = null;
    const idleTimeoutSeconds = idleTimeoutMs / 1000;
    // Periodic idle warning state
    let warningHandle = null;
    let idleMinuteCounter = 0;
    const startWarningInterval = () => {
        if (warningHandle !== null)
            clearInterval(warningHandle);
        idleMinuteCounter = 0;
        warningHandle = setInterval(() => {
            idleMinuteCounter++;
            onIdleWarning(idleMinuteCounter);
        }, idleWarningIntervalMs);
    };
    const resetIdleTimer = () => {
        if (timeoutHandle !== null)
            clearTimeout(timeoutHandle);
        timeoutHandle = setTimeout(() => {
            Effect.runPromise(Deferred.fail(timeoutSignal, new TimeoutError({
                message: `Agent idle for ${idleTimeoutSeconds} seconds — no output received. Consider increasing the idle timeout with --idle-timeout.`,
                idleTimeoutSeconds,
            }))).catch(() => { });
        }, idleTimeoutMs);
        // Reset warning interval on activity
        startWarningInterval();
    };
    resetIdleTimer();
    const execEffect = Effect.gen(function* () {
        const execResult = yield* sandbox.execStreaming(provider.buildPrintCommand(prompt), (line) => {
            for (const parsed of provider.parseStreamLine(line)) {
                if (parsed.type === "text") {
                    resetIdleTimer();
                    onText(parsed.text);
                }
                else if (parsed.type === "result") {
                    resultText = parsed.result;
                }
                else if (parsed.type === "tool_call") {
                    resetIdleTimer();
                    onToolCall(parsed.name, parsed.args);
                }
            }
        }, { cwd: sandboxRepoDir });
        if (execResult.exitCode !== 0) {
            return yield* Effect.fail(new AgentError({
                message: `${provider.name} exited with code ${execResult.exitCode}:\n${execResult.stderr}`,
            }));
        }
        return { result: resultText || execResult.stdout };
    }).pipe(Effect.ensuring(Effect.sync(() => {
        if (timeoutHandle !== null) {
            clearTimeout(timeoutHandle);
            timeoutHandle = null;
        }
        if (warningHandle !== null) {
            clearInterval(warningHandle);
            warningHandle = null;
        }
    })));
    return yield* Effect.raceFirst(execEffect, Deferred.await(timeoutSignal));
});
const DEFAULT_COMPLETION_SIGNAL = "<promise>COMPLETE</promise>";
const DEFAULT_IDLE_TIMEOUT_SECONDS = 10 * 60; // 600 seconds
export const orchestrate = (options) => {
    const idleTimeoutMs = (options.idleTimeoutSeconds ?? DEFAULT_IDLE_TIMEOUT_SECONDS) * 1000;
    return Effect.gen(function* () {
        const factory = yield* SandboxFactory;
        const display = yield* Display;
        const { hostRepoDir, sandboxRepoDir, iterations, hooks, prompt, branch, provider, } = options;
        let completionSignals;
        if (options.completionSignal === undefined) {
            completionSignals = [DEFAULT_COMPLETION_SIGNAL];
        }
        else if (Array.isArray(options.completionSignal)) {
            completionSignals = options.completionSignal;
        }
        else {
            completionSignals = [options.completionSignal];
        }
        const label = (msg) => options.name ? `[${options.name}] ${msg}` : msg;
        const allCommits = [];
        let allStdout = "";
        let resolvedBranch = "";
        let iterationPreservedPath;
        for (let i = 1; i <= iterations; i++) {
            yield* display.status(label(`Iteration ${i}/${iterations}`), "info");
            const sandboxResult = yield* factory.withSandbox(({ hostWorktreePath }) => withSandboxLifecycle({
                hostRepoDir,
                sandboxRepoDir,
                hooks,
                branch,
                hostWorktreePath,
            }, (ctx) => Effect.gen(function* () {
                // Preprocess prompt (run !`command` expressions inside sandbox)
                const fullPrompt = yield* preprocessPrompt(prompt, ctx.sandbox, ctx.sandboxRepoDir);
                yield* display.status(label("Agent started"), "success");
                // Invoke the agent
                const onText = (text) => {
                    Effect.runPromise(display.text(text));
                };
                const onToolCall = (name, formattedArgs) => {
                    Effect.runPromise(display.toolCall(name, formattedArgs));
                };
                const onIdleWarning = (minutes) => {
                    const msg = minutes === 1
                        ? "Agent idle for 1 minute"
                        : `Agent idle for ${minutes} minutes`;
                    Effect.runPromise(display.status(label(msg), "warn"));
                };
                const { result: agentOutput } = yield* invokeAgent(ctx.sandbox, ctx.sandboxRepoDir, fullPrompt, provider, idleTimeoutMs, onText, onToolCall, onIdleWarning, options._idleWarningIntervalMs);
                yield* display.status(label("Agent stopped"), "info");
                // Check completion signal
                const matchedSignal = completionSignals.find((sig) => agentOutput.includes(sig));
                return {
                    completionSignal: matchedSignal,
                    stdout: agentOutput,
                };
            })));
            const lifecycleResult = sandboxResult.value;
            iterationPreservedPath = sandboxResult.preservedWorktreePath;
            allCommits.push(...lifecycleResult.commits);
            allStdout += lifecycleResult.result.stdout;
            resolvedBranch = lifecycleResult.branch;
            if (lifecycleResult.result.completionSignal !== undefined) {
                yield* display.status(label(`Agent signaled completion after ${i} iteration(s).`), "success");
                return {
                    iterationsRun: i,
                    completionSignal: lifecycleResult.result.completionSignal,
                    stdout: allStdout,
                    commits: allCommits,
                    branch: resolvedBranch,
                    preservedWorktreePath: iterationPreservedPath,
                };
            }
        }
        yield* display.status(label(`Reached max iterations (${iterations}).`), "info");
        return {
            iterationsRun: iterations,
            completionSignal: undefined,
            stdout: allStdout,
            commits: allCommits,
            branch: resolvedBranch,
            preservedWorktreePath: iterationPreservedPath,
        };
    });
};
