// @ai-hero/sandcastle@0.4.0/dist/run.js
import { NodeContext, NodeFileSystem } from "@effect/platform-node";
import path, { join } from "node:path";
import { styleText } from "node:util";
import { Effect, Layer } from "effect";
import { ClackDisplay, Display, FileDisplay, } from "./Display.js";
import { orchestrate } from "./Orchestrator.js";
import { resolvePrompt } from "./PromptResolver.js";
import { WorktreeDockerSandboxFactory, SandboxConfig, SANDBOX_WORKSPACE_DIR, } from "./SandboxFactory.js";
import { resolveEnv } from "./EnvResolver.js";
import { generateTempBranchName, getCurrentBranch } from "./WorktreeManager.js";
import { substitutePromptArgs, validateNoBuiltInArgOverride, BUILT_IN_PROMPT_ARG_KEYS, } from "./PromptArgumentSubstitution.js";
/** Default maximum number of iterations for a run. */
export const DEFAULT_MAX_ITERATIONS = 1;
/** Replace characters that are invalid or problematic in file paths with dashes. */
export const sanitizeBranchForFilename = (branch) => branch.replace(/[/\\:*?"<>|]/g, "-");
/**
 * Print the startup message to the terminal when using file-based logging.
 * Uses styleText for lightweight bold/dim styling — does not use Clack.
 */
export const printFileDisplayStartup = (options) => {
    const name = options.agentName ?? "Agent";
    const label = styleText("bold", `[${name}]`);
    const branchPart = options.branch ? ` on branch ${options.branch}` : "";
    const relativeLogPath = path.relative(process.cwd(), options.logPath);
    console.log(`${label} Started${branchPart}`);
    console.log(styleText("dim", `  tail -f ${relativeLogPath}`));
};
/**
 * Build the log filename for a run.
 * When a targetBranch is provided (temp branch mode), prefixes the filename
 * with the sanitized target branch name so developers can identify which
 * branch the run was targeting: `<targetBranch>-<resolvedBranch>.log`
 * When no targetBranch, uses just the resolved branch: `<resolvedBranch>.log`
 * When a name is provided, appends it to avoid collisions in multi-agent workflows.
 */
export const buildLogFilename = (resolvedBranch, targetBranch, name) => {
    const sanitized = sanitizeBranchForFilename(resolvedBranch);
    const nameSuffix = name
        ? `-${name.toLowerCase().replace(/[^a-z0-9_.-]/g, "-")}`
        : "";
    if (targetBranch) {
        return `${sanitizeBranchForFilename(targetBranch)}-${sanitized}${nameSuffix}.log`;
    }
    return `${sanitized}${nameSuffix}.log`;
};
/**
 * Build the summary rows for a run, used in both terminal mode and
 * log-to-file mode. When a custom name is provided it appears as the
 * Agent value instead of the internal provider name.
 */
export const buildRunSummaryRows = (options) => ({
    Agent: options.name ?? options.agentName,
    Sandbox: options.sandboxName,
    "Max iterations": String(options.maxIterations),
    Branch: options.branch,
});
/**
 * Build the completion status message for a run, used in both terminal mode
 * and log-to-file mode to record the final outcome.
 */
export const buildCompletionMessage = (completionSignal, iterationsRun) => {
    if (completionSignal !== undefined) {
        return {
            message: `Run complete: agent finished after ${iterationsRun} iteration(s).`,
            severity: "success",
        };
    }
    return {
        message: `Run complete: reached ${iterationsRun} iteration(s) without completion signal.`,
        severity: "warn",
    };
};
export const run = async (options) => {
    const { prompt, promptFile, maxIterations = DEFAULT_MAX_ITERATIONS, hooks, agent: provider, } = options;
    // Derive branch strategy from the sandbox provider
    const branchStrategy = options.sandbox.branchStrategy;
    const effectiveBranchType = branchStrategy.type;
    // Validate: copyToSandbox is incompatible with head strategy
    if (effectiveBranchType === "head" &&
        options.copyToSandbox &&
        options.copyToSandbox.length > 0) {
        throw new Error("copyToSandbox is not supported with head branch strategy. " +
            "In head mode the host working directory is bind-mounted directly.");
    }
    // Extract explicit branch when in branch mode
    const branch = branchStrategy.type === "branch" ? branchStrategy.branch : undefined;
    const hostRepoDir = process.cwd();
    // Resolve prompt
    const rawPrompt = await Effect.runPromise(resolvePrompt({ prompt, promptFile }).pipe(Effect.provide(NodeContext.layer)));
    const agentName = provider.name;
    // Resolve env vars
    const env = await Effect.runPromise(resolveEnv(hostRepoDir).pipe(Effect.provide(NodeContext.layer)));
    // Always capture the host's current branch for the TARGET_BRANCH built-in
    // prompt argument. When using a temp branch, it also prefixes the log filename.
    const currentHostBranch = await Effect.runPromise(getCurrentBranch(hostRepoDir));
    // When in merge-to-head mode, generate a temporary branch name.
    // In head mode, use the host's current branch directly (no worktree).
    const resolvedBranch = effectiveBranchType === "head"
        ? currentHostBranch
        : (branch ?? generateTempBranchName(options.name));
    // When using a temp branch, prefix the log filename with the target branch
    // (the host's current branch) so developers can tell which branch was targeted.
    const targetBranch = effectiveBranchType === "merge-to-head" ? currentHostBranch : undefined;
    // Resolve logging option
    const resolvedLogging = options.logging ?? {
        type: "file",
        path: join(hostRepoDir, ".sandcastle", "logs", buildLogFilename(resolvedBranch, targetBranch, options.name)),
    };
    const displayLayer = resolvedLogging.type === "file"
        ? (() => {
            printFileDisplayStartup({
                logPath: resolvedLogging.path,
                agentName: options.name,
                branch: resolvedBranch,
            });
            return Layer.provide(FileDisplay.layer(resolvedLogging.path), NodeFileSystem.layer);
        })()
        : ClackDisplay.layer;
    const factoryLayer = Layer.provide(WorktreeDockerSandboxFactory.layer, Layer.mergeAll(Layer.succeed(SandboxConfig, {
        env,
        hostRepoDir,
        copyToSandbox: options.copyToSandbox,
        name: options.name,
        sandboxProvider: options.sandbox,
    }), NodeFileSystem.layer, displayLayer));
    const runLayer = Layer.merge(factoryLayer, displayLayer);
    const result = await Effect.runPromise(Effect.gen(function* () {
        const d = yield* Display;
        yield* d.intro(options.name ?? "sandcastle");
        const rows = buildRunSummaryRows({
            name: options.name,
            agentName,
            sandboxName: options.sandbox.name,
            maxIterations,
            branch: resolvedBranch,
        });
        yield* d.summary("Sandcastle Run", rows);
        // Validate that the user has not provided built-in prompt argument keys
        const userArgs = options.promptArgs ?? {};
        yield* validateNoBuiltInArgOverride(userArgs);
        // Build effective args: built-in args merged with user-provided args.
        // In none mode, resolvedBranch is already currentHostBranch, so
        // SOURCE_BRANCH and TARGET_BRANCH both resolve to the host's current branch.
        const effectiveArgs = {
            SOURCE_BRANCH: resolvedBranch,
            TARGET_BRANCH: currentHostBranch,
            ...userArgs,
        };
        const builtInArgKeysSet = new Set(BUILT_IN_PROMPT_ARG_KEYS);
        const resolvedPrompt = yield* substitutePromptArgs(rawPrompt, effectiveArgs, builtInArgKeysSet);
        // In head mode, pass the host branch so SandboxLifecycle skips the merge step.
        // In merge-to-head mode, branch is undefined (triggers merge). In branch mode, it's the explicit branch.
        const orchestrateBranch = effectiveBranchType === "head" ? currentHostBranch : branch;
        const orchestrateResult = yield* orchestrate({
            hostRepoDir,
            sandboxRepoDir: SANDBOX_WORKSPACE_DIR,
            iterations: maxIterations,
            hooks,
            prompt: resolvedPrompt,
            branch: orchestrateBranch,
            provider,
            completionSignal: options.completionSignal,
            idleTimeoutSeconds: options.idleTimeoutSeconds,
            name: options.name,
        });
        const completion = buildCompletionMessage(orchestrateResult.completionSignal, orchestrateResult.iterationsRun);
        yield* d.status(completion.message, completion.severity);
        return orchestrateResult;
    }).pipe(Effect.provide(runLayer)));
    return {
        ...result,
        logFilePath: resolvedLogging.type === "file" ? resolvedLogging.path : undefined,
    };
};
