// @ai-hero/sandcastle@0.4.0/dist/SandboxFactory.js
import { Context, Effect, Exit, Layer } from "effect";
import { FileSystem } from "@effect/platform";
import { existsSync } from "node:fs";
import { join, resolve } from "node:path";
import { AgentError, CopyError, ExecError, TimeoutError, WorktreeError, } from "./errors.js";
import * as WorktreeManager from "./WorktreeManager.js";
import { copyToSandbox } from "./CopyToSandbox.js";
import { Display } from "./Display.js";
import { syncIn } from "./syncIn.js";
import { syncOut } from "./syncOut.js";
export class Sandbox extends Context.Tag("Sandbox")() {
}
/**
 * Wrap a Promise-based sandbox handle into an Effect-based SandboxService layer.
 * Works with both bind-mount handles (copyIn/copyOut unsupported) and
 * isolated handles (copyIn/copyOut delegated to the handle).
 */
export const makeSandboxLayerFromHandle = (handle) => Layer.succeed(Sandbox, {
    exec: (command, options) => Effect.tryPromise({
        try: () => handle.exec(command, options),
        catch: (e) => new ExecError({
            command,
            message: `exec failed: ${e instanceof Error ? e.message : String(e)}`,
        }),
    }),
    execStreaming: (command, onStdoutLine, options) => Effect.tryPromise({
        try: () => handle.execStreaming(command, onStdoutLine, options),
        catch: (e) => new ExecError({
            command,
            message: `exec streaming failed: ${e instanceof Error ? e.message : String(e)}`,
        }),
    }),
    copyIn: "copyIn" in handle
        ? (hostPath, sandboxPath) => Effect.tryPromise({
            try: () => handle.copyIn(hostPath, sandboxPath),
            catch: (e) => new CopyError({
                message: `copyIn failed: ${e instanceof Error ? e.message : String(e)}`,
            }),
        })
        : () => Effect.fail(new CopyError({
            message: "copyIn is not supported for bind-mount sandbox providers",
        })),
    copyOut: "copyOut" in handle
        ? (sandboxPath, hostPath) => Effect.tryPromise({
            try: () => handle.copyOut(sandboxPath, hostPath),
            catch: (e) => new CopyError({
                message: `copyOut failed: ${e instanceof Error ? e.message : String(e)}`,
            }),
        })
        : () => Effect.fail(new CopyError({
            message: "copyOut is not supported for bind-mount sandbox providers",
        })),
});
/** The mount point inside the container where the project worktree is bound. */
export const SANDBOX_WORKSPACE_DIR = "/home/agent/workspace";
export class SandboxFactory extends Context.Tag("SandboxFactory")() {
}
export class SandboxConfig extends Context.Tag("SandboxConfig")() {
}
/** @deprecated Use SandboxConfig instead. */
export const WorktreeSandboxConfig = SandboxConfig;
/**
 * Print a message to stderr about a preserved worktree, with review and cleanup instructions.
 */
const printWorktreePreservedMessage = (worktreePath, reason) => {
    console.error(`\n${reason}`);
    console.error(`  To review: cd ${worktreePath}`);
    console.error(`  To clean up: git worktree remove --force ${worktreePath}`);
};
/**
 * Resolves the git-related mounts needed for the sandbox.
 * Handles both normal repos (where .git is a directory) and worktrees
 * (where .git is a file pointing to the parent repo's .git/worktrees/<name>).
 */
export const resolveGitMounts = (gitPath) => Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;
    const stat = yield* fs.stat(gitPath);
    if (stat.type === "Directory") {
        return [{ hostPath: gitPath, sandboxPath: gitPath }];
    }
    // Worktree: .git is a file with "gitdir: <path>"
    const content = (yield* fs.readFileString(gitPath)).trim();
    const match = content.match(/^gitdir:\s*(.+)$/);
    if (!match) {
        // Unrecognized format — fall back to mounting the file as-is
        return [{ hostPath: gitPath, sandboxPath: gitPath }];
    }
    const gitdirPath = match[1];
    // gitdirPath is like /path/to/repo/.git/worktrees/<name>
    // Mount both the .git file and the parent .git directory
    const parentGitDir = resolve(gitdirPath, "..", "..");
    return [
        { hostPath: gitPath, sandboxPath: gitPath },
        { hostPath: parentGitDir, sandboxPath: parentGitDir },
    ];
});
/**
 * Start a sandbox using the provider abstraction.
 * Returns the handle, sandbox layer, and workspace path.
 */
const startProviderSandbox = (provider, worktreeOrRepoPath, hostRepoDir, env, gitMounts, workspaceDir) => Effect.tryPromise({
    try: () => {
        const mounts = [
            {
                hostPath: worktreeOrRepoPath,
                sandboxPath: workspaceDir,
            },
            ...gitMounts,
        ];
        return provider.create({
            worktreePath: worktreeOrRepoPath,
            hostRepoPath: hostRepoDir,
            mounts,
            env,
        });
    },
    catch: (e) => new WorktreeError({
        message: `Provider '${provider.name}' create failed: ${e instanceof Error ? e.message : String(e)}`,
    }),
}).pipe(Effect.map((handle) => ({
    handle,
    sandboxLayer: makeSandboxLayerFromHandle(handle),
    workspacePath: handle.workspacePath,
})));
/**
 * Start an isolated sandbox: create handle, sync host repo via git bundle.
 * Returns the handle, sandbox layer, and workspace path.
 */
const startIsolatedProviderSandbox = (provider, hostRepoDir, env, copyPaths) => Effect.gen(function* () {
    const handle = yield* Effect.tryPromise({
        try: () => provider.create({ env }),
        catch: (e) => new WorktreeError({
            message: `Isolated provider '${provider.name}' setup failed: ${e instanceof Error ? e.message : String(e)}`,
        }),
    });
    yield* syncIn(hostRepoDir, handle);
    if (copyPaths && copyPaths.length > 0) {
        for (const relativePath of copyPaths) {
            const hostPath = join(hostRepoDir, relativePath);
            if (!existsSync(hostPath)) {
                continue;
            }
            const sandboxPath = join(handle.workspacePath, relativePath);
            yield* Effect.tryPromise({
                try: () => handle.copyIn(hostPath, sandboxPath),
                catch: (e) => new WorktreeError({
                    message: `Failed to copy ${relativePath} into sandbox: ${e instanceof Error ? e.message : String(e)}`,
                }),
            });
        }
    }
    return {
        handle,
        sandboxLayer: makeSandboxLayerFromHandle(handle),
        workspacePath: handle.workspacePath,
    };
});
export const WorktreeDockerSandboxFactory = {
    layer: Layer.effect(SandboxFactory, Effect.gen(function* () {
        const { env, hostRepoDir, copyToSandbox: copyPaths, name, sandboxProvider, } = yield* SandboxConfig;
        // Read branch strategy from the provider
        const branchStrategy = sandboxProvider.branchStrategy;
        const isHeadMode = branchStrategy.type === "head";
        const branch = branchStrategy.type === "branch" ? branchStrategy.branch : undefined;
        const fileSystem = yield* FileSystem.FileSystem;
        const display = yield* Display;
        return {
            withSandbox: (makeEffect) => {
                // Isolated providers: skip worktree, sync via git bundle
                if (sandboxProvider.tag === "isolated") {
                    return Effect.acquireUseRelease(startIsolatedProviderSandbox(sandboxProvider, hostRepoDir, env, copyPaths), 
                    // Use
                    ({ sandboxLayer }) => makeEffect({}).pipe(Effect.provide(sandboxLayer)), 
                    // Release: sync commits back to host, then close
                    ({ handle }) => syncOut(hostRepoDir, handle).pipe(Effect.catchAll((e) => Effect.sync(() => {
                        console.error(`[sandcastle] Warning: syncOut failed: ${e.message}`);
                    })), Effect.andThen(Effect.tryPromise({
                        try: () => handle.close(),
                        catch: () => undefined,
                    })), Effect.orDie)).pipe(Effect.map((value) => ({
                        value,
                        preservedWorktreePath: undefined,
                    })));
                }
                if (isHeadMode) {
                    // Head mode: bind-mount host directory directly, no worktree
                    const gitPath = join(hostRepoDir, ".git");
                    return resolveGitMounts(gitPath).pipe(Effect.provideService(FileSystem.FileSystem, fileSystem), Effect.mapError((e) => new WorktreeError({
                        message: `Failed to resolve git mounts: ${e}`,
                    })), Effect.flatMap((gitMounts) => Effect.acquireUseRelease(startProviderSandbox(sandboxProvider, hostRepoDir, hostRepoDir, env, gitMounts, SANDBOX_WORKSPACE_DIR), 
                    // Use
                    ({ sandboxLayer }) => makeEffect({}).pipe(Effect.provide(sandboxLayer)), 
                    // Release
                    ({ handle }) => Effect.tryPromise({
                        try: () => handle.close(),
                        catch: () => undefined,
                    }).pipe(Effect.orDie)).pipe(Effect.map((value) => ({
                        value,
                        preservedWorktreePath: undefined,
                    })))));
                }
                // Worktree mode (merge-to-head or explicit branch)
                // Populated by the release phase when a worktree is preserved on failure,
                // so we can attach the path to recognized error types before they propagate.
                let preservedWorktreePath;
                return Effect.acquireUseRelease(
                // Acquire: prune stale worktrees (best-effort), create worktree, then start sandbox
                WorktreeManager.pruneStale(hostRepoDir)
                    .pipe(Effect.catchAll((e) => Effect.sync(() => {
                    console.error("[sandcastle] Warning: failed to prune stale worktrees:", e.message);
                })))
                    .pipe(Effect.andThen(branch
                    ? WorktreeManager.create(hostRepoDir, { branch })
                    : WorktreeManager.create(hostRepoDir, { name })))
                    .pipe(Effect.provideService(FileSystem.FileSystem, fileSystem))
                    .pipe(Effect.flatMap((worktreeInfo) => (copyPaths && copyPaths.length > 0
                    ? display.spinner("Copying to sandbox", copyToSandbox(copyPaths, hostRepoDir, worktreeInfo.path))
                    : Effect.succeed(undefined)).pipe(Effect.map(() => worktreeInfo))))
                    .pipe(Effect.flatMap((worktreeInfo) => {
                    const gitPath = join(hostRepoDir, ".git");
                    return resolveGitMounts(gitPath).pipe(Effect.provideService(FileSystem.FileSystem, fileSystem), Effect.mapError((e) => new WorktreeError({
                        message: `Failed to resolve git mounts: ${e}`,
                    })), Effect.flatMap((gitMounts) => 
                    // sandboxProvider is guaranteed bind-mount here
                    // (isolated providers return early above)
                    startProviderSandbox(sandboxProvider, worktreeInfo.path, hostRepoDir, env, gitMounts, SANDBOX_WORKSPACE_DIR).pipe(Effect.map(({ handle, sandboxLayer }) => ({
                        worktreeInfo,
                        handle,
                        sandboxLayer,
                    })))));
                })), 
                // Use
                ({ worktreeInfo, sandboxLayer }) => makeEffect({ hostWorktreePath: worktreeInfo.path }).pipe(Effect.provide(sandboxLayer)), 
                // Release: close provider handle, then remove/preserve worktree based on dirty state.
                ({ worktreeInfo, handle }, exit) => Effect.tryPromise({
                    try: () => handle.close(),
                    catch: () => undefined,
                }).pipe(Effect.asVoid, Effect.andThen(WorktreeManager.hasUncommittedChanges(worktreeInfo.path).pipe(Effect.catchAll(() => Effect.succeed(false)), Effect.flatMap((isDirty) => {
                    if (isDirty) {
                        preservedWorktreePath = worktreeInfo.path;
                        printWorktreePreservedMessage(worktreeInfo.path, Exit.isSuccess(exit)
                            ? `Run succeeded but worktree has uncommitted changes at ${worktreeInfo.path}`
                            : `Worktree preserved at ${worktreeInfo.path}`);
                        return Effect.void;
                    }
                    else {
                        if (!Exit.isSuccess(exit)) {
                            console.error(`\nWorktree removed (no uncommitted changes)`);
                        }
                        return WorktreeManager.remove(worktreeInfo.path);
                    }
                }))), Effect.orDie)).pipe(Effect.map((value) => ({
                    value,
                    preservedWorktreePath,
                })), 
                // Attach the preserved worktree path to TimeoutError and AgentError so
                // programmatic callers can build on top of the preserved worktree.
                Effect.mapError((e) => {
                    const path = preservedWorktreePath;
                    if (path !== undefined) {
                        if (e instanceof TimeoutError) {
                            return new TimeoutError({
                                message: e.message,
                                idleTimeoutSeconds: e.idleTimeoutSeconds,
                                preservedWorktreePath: path,
                            });
                        }
                        if (e instanceof AgentError) {
                            return new AgentError({
                                message: e.message,
                                preservedWorktreePath: path,
                            });
                        }
                    }
                    return e;
                }));
            },
        };
    })),
};
